oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -146,66 +129,230 @@ impl GrpcServerNode {
-    fn decode_grpc_request(
-        http_request_path: &str,
-        http_request_body: &dyn hyper::body::Buf,
-    ) -> Result<GrpcRequest, GrpcServerError> {
-        // Parse an HTTP request body as an [`Any`] message.
-        let grpc_request_body = Any::decode(http_request_body.bytes()).map_err(|error| {
-            error!("Failed to parse Protobuf message {}", error);
-            GrpcServerError::BadProtobufMessage
-        })?;
+/// Oak Node implementation for the gRPC server.
+impl super::Node for GrpcServerNode {
+    fn start(self: Box<Self>) -> Result<JoinHandle<()>, OakStatus> {
+        let thread_handle = thread::Builder::new()
+            .name(self.to_string())
+            .spawn(move || self.worker_thread())
+            .expect("Failed to spawn thread");
+        Ok(thread_handle)
+    }
+}
 
-        // Create a gRPC request.
-        encap_request(&grpc_request_body, http_request_path).ok_or_else(|| {
-            error!("Failed to create a GrpcRequest");
-            GrpcServerError::BadProtobufMessage
-        })
+/// [`HttpRequestHandler`] handles HTTP/2 requests from a client and sends HTTP/2 responses back.
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to a Runtime that corresponds to a node that created a gRPC server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing gRPC invocations.
+    writer: Handle,
+}
+
+// Set a mandatory prefix for all gRPC requests processed by a gRPC pseudo-Node.
+impl NamedService for HttpRequestHandler {
+    const NAME: &'static str = "oak";
+}
+
+impl Service<http::Request<hyper::Body>> for HttpRequestHandler {
+    type Response = http::Response<tonic::body::BoxBody>;
+    type Error = http::Error;
+    type Future = BoxFuture<Self::Response, Self::Error>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
     }
 
-    /// Creates an HTTP response message.
-    fn http_response(status: http::StatusCode, body: Vec<u8>) -> hyper::Response<hyper::Body> {
-        let mut response = hyper::Response::new(hyper::Body::from(body));
-        *response.status_mut() = status;
-        response
+    // Decodes an unary gRPC request using a [`VecCodec`] and processes it with
+    // [`tonic::server::Grpc::unary`] and a [`GrpcRequestHandler`].
+    fn call(&mut self, request: http::Request<hyper::Body>) -> Self::Future {
+        let grpc_handler = GrpcRequestHandler::new(
+            self.runtime.clone(),
+            self.writer,
+            request.uri().path().to_string(),
+        );
+
+        let future = async move {
+            debug!("Processing an HTTP/2 request: {:?}", request);
+            let mut grpc_service = Grpc::new(VecCodec::default());
+            let response = grpc_service.unary(grpc_handler, request).await;
+            debug!("Sending an HTTP/2 response: {:?}", response);
+            Ok(response)
+        };
+
+        Box::pin(future)
```

Comment:
I am not entirely sure why all this needs to happen, vs just getting the bytes out of the `http::Request`. Are you sure we need all these extra abstractions that ultimately don't do anything (apart from converting between different formats for bytes)?

Reply:
I agree. I think it should be possible to just use `hyper::body::aggregate` to get the bytes from the request body, rather than using a codec and a grpc service.

Reply:
Fine for now, but note that we will probably end up with a HTTP server pseudo-node, so this will end up being basically that, and that logic will probably be pushed into the SDK.